### PR TITLE
feat: Add rados namespace support to rbd plugin

### PR DESCRIPTION
**Pull request by mehdy**
# Describe what this PR does #

Make sure to operate within the namespace if any given when dealing
with rbd images and snapshots and volume and snapshot journals.

Add some basic e2e tests using a namespace.

## Is there anything that requires special attention ##

Do you have any questions?
No.

Is the change backward compatible?
Yes.

Are there concerns around backward compatibility?
No.

Provide any external context for the change, if any.

* [rbd namespace](https://docs.ceph.com/docs/master/man/8/rbd/?highlight=namespace)

## Related issues ##

Mention any github issues relevant to this PR. Adding below line
will help to auto close the issue once the PR is merged.

Fixes: #798 

## Future concerns ##

List items that are not part of the PR and do not impact it's
functionality, but are work items that can be taken up subsequently.

### CODE REVIEW

**Comment by mehdy** [comment]
I'm not sure why the **Semantic Pull Request** bot is failing. I tried to write the commit message and PR title similar to other commits and PRs. I'd be glad to get a hint if it needs more editing.

**Comment by mehdy** [comment]
I'm using `SetNamespace` method a lot in this PR and it was removed this [commit](https://github.com/ceph/ceph-csi/commit/f091374ba9eb03d53d51a777f2803abb6c83f1b6) :))
Should I use `journal.namespace = ns` or rewrite a new `SetNamespace` method again or would you revert this commit?
@Madhu-1

**Comment by Madhu-1** [comment]
> I'm using `SetNamespace` method a lot in this PR and it was removed this [commit](https://github.com/ceph/ceph-csi/commit/f091374ba9eb03d53d51a777f2803abb6c83f1b6) :))
> Should I use `journal.namespace = ns` or rewrite a new `SetNamespace` method again or would you revert this commit?
> @Madhu-1

having `SetNamespace` makes sense now as you need to call it multiple times. @phlogistonjohn any thoughts?

**Comment by phlogistonjohn** [comment]
Mutating a global object frequently strikes me as a "code smell" that should be avoided in general. That's one of the reasons I removed it. 

I quickly looked at your PR here and see a very common pattern, and since I'm also attempting to refactor how the journal code works, many call sites I've become familiar with over the last few days. In these cases you're calling SetNamespace right before a call to something like volJournal.CheckReservation.

Without digging deeper into your changes, I'd suggest altering the journal function arguments and adding namespace there rather than repeatedly mutating the state of the volJournal/etc.

FWIW, I'm also trying to break up the functionality such that the operational functions in voljournal.go rely more on local state via a new type rather than the global volJournal/snapJournal values. This means I too will be changing the arguments of the same functions. So this is just a heads up that we're likely to make merge conflicts with each other. I'm not trying to worry or discourage you... just making you aware of it. :-)

**Comment by mehdy** [comment]
@phlogistonjohn I'll change the journal functions to receive namespace as an argument as you suggested. Thanks!

About the refactoring work you mentioned, wouldn't it be better if you do it after this PR? :D

**Comment by mehdy** [comment]
Well, now the PR is ready to be merged.

I'll explain the reasons for the changes I made:

### Changing Travis CI dist to bionic
Since the e2e tests are using krbd and xenial's rbd kernel module doesn't support namespace, we need a newer version of Linux, hence the bionic release.
But we still need xenial for arm64 builds.
The realpath package is now included in coreutils package in xenial.


### New e2e tests added to Travis CI
Re-running tests for the namespace within the same job fails because the old test job changes the clean state of kubernetes and ceph (not sure which and why).
I think running the namespace tests concurrently as an independent job is cleaner and speeds up the whole CI process as well.

### The extra kubeket.resolv-conf
Bionic uses systems-resolve and coreDNS will be stuck in a loop when using the default /etc/resolv.conf. I solved it using [this solution](https://github.com/kubernetes/minikube/issues/3511#issuecomment-461664952)

@Madhu-1

**Comment by Madhu-1** [comment]
@mehdy thanks , I will take a look at it.

File: internal/rbd/rbd_util.go [hunk 1]
```diff
@@ -954,13 +1037,23 @@ func stashRBDImageMetadata(volOptions *rbdVolume, path string) error {
 
 	encodedBytes, err := json.Marshal(imgMeta)
 	if err != nil {
+		if volOptions.Namespace != "" {
```
Comment (Madhu-1) [inline internal/rbd/rbd_util.go:1040]: construct the error and return once

**Comment by Madhu-1** [comment]
@dillaman PTAL

File: internal/rbd/nodeserver.go [hunk 2]
```diff
@@ -687,7 +687,7 @@ func getDevicePath(ctx context.Context, volumePath string) (string, error) {
 	if err != nil {
 		klog.Errorf(util.Log(ctx, "failed to find image metadata: %v"), err)
 	}
-	device, found := findDeviceMappingImage(ctx, imgInfo.Pool, imgInfo.ImageName, imgInfo.NbdAccess)
+	device, found := findDeviceMappingImage(ctx, imgInfo.Pool, imgInfo.Namespace, imgInfo.ImageName, imgInfo.NbdAccess)
```
Comment (nixpanic) [inline internal/rbd/nodeserver.go:690]: TODO: `findDeviceMappingImage()` should probably become `imgInfo.findDeviceMappingImage()`

This can be placed in its own commit, or even its own PR

File: internal/rbd/rbd_util.go [hunk 3]
```diff
@@ -156,6 +158,13 @@ func (rv *rbdVolume) Destroy() {
 	}
 }
 
+func getImageSpec(pool, namespace, imageName string) string {
```
Comment (nixpanic) [inline internal/rbd/rbd_util.go:161]: can this be made a method for the `rbdVolume` type?

File: scripts/minikube.sh [hunk 4]
```diff
@@ -78,7 +78,8 @@ fi
 K8S_FEATURE_GATES=${K8S_FEATURE_GATES:-"BlockVolume=true,CSIBlockVolume=true,VolumeSnapshotDataSource=true,ExpandCSIVolumes=true"}
 
 #extra-config for kube https://minikube.sigs.k8s.io/docs/reference/configuration/kubernetes/
-EXTRA_CONFIG=${EXTRA_CONFIG:-"--extra-config=apiserver.enable-admission-plugins=PodSecurityPolicy"}
+EXTRA_CONFIG=${EXTRA_CONFIG:-"--extra-config=apiserver.enable-admission-plugins=PodSecurityPolicy \
```
Comment (nixpanic) [inline scripts/minikube.sh:81]: this should probably be its own commit, or part of some other CI change

**Comment by nixpanic** [comment]
> I really like the `getImageSpec()` function to be split into its own commit. You can send it even as a PR without the namespace change as a preparation. It will cleanup this PR a lot, as many of the changes are related to it. With this PR you can then enhance `getImageSpec()` to include the namespace. This will make reviewing much easier.

If you want, I can send a PR for this to save you some time. Just let me know if you want me to take care of that.

**Comment by mehdy** [comment]
@Madhu-1 @nixpanic Thank you both. These are really good points and I'll change the PR as you advised.

File: internal/rbd/nodeserver.go (hunk 2 above)
Comment (mehdy) [reply internal/rbd/nodeserver.go:696]: `rbdVolume` uses the `findDeviceMappingImage` [here](https://github.com/ceph/ceph-csi/pull/1035/files#diff-9e4cf5c8370e4beee82e84b9818a53acL146).

should I implement it for both of them or just leave it like this?

**Comment by nixpanic** [comment]
Hi @mehdy , I think the preparations for this have now all been merged. Can you rebase this PR and force-push it again?

Thanks!

**Comment by mehdy** [comment]
@nixpanic sure will do! I just need a little more time to write specific e2e tests.

File: internal/rbd/nodeserver.go (hunk 2 above)
Comment (nixpanic) [reply internal/rbd/nodeserver.go:697]: Not needed for now, it is something that might cleanup the code and improve maintenance over time.

File: internal/util/csiconfig.go [hunk 5]
```diff
@@ -46,6 +48,7 @@ type ClusterInfo struct {
 // [
 // 	{
 // 		"clusterID": "<cluster-id>",
+//		"namespace": "<namespace>",
```
Comment (nixpanic) [inline internal/util/csiconfig.go:51]: use `"radosNamespace"` in the comment too please

Comment (mehdy) [reply internal/util/csiconfig.go]: Sorry, I missed it :)) it's now fixed.

**Comment by mehdy** [comment]
@nixpanic This is ready now. Could you please take a look at it?
Thanks.

**Comment by mehdy** [comment]
BUMP.
@Madhu-1 @nixpanic

File: internal/rbd/rbd_util.go [hunk 6]
```diff
@@ -230,6 +241,7 @@ func (rv *rbdVolume) openIoctx() error {
 		return err
 	}
 
+	ioctx.SetNamespace(pOpts.RadosNamespace)
```
Comment (thojkooi) [inline internal/rbd/rbd_util.go:244]: I think this needs to be 
```go
ioctx.SetNamespace(rv.RadosNamespace)
```

Currently the build fails because `pOpts` is undefined in this function.

**Comment by mehdy** [comment]
All tests have passed and the code is ready to be reviewed :)
@Madhu-1 @nixpanic

**Comment by Madhu-1** [comment]
Thanks for completing this one. in ceph-csi meeting, we have decided to review this PR after some time as we have pending/ongoing PR's to fix P0 issues which we have planned for release-v3.0.0.

**Comment by mehdy** [comment]
@Madhu-1 It's ok! thanks. BTW if there were any other issues I could help you with, feel free to contact me. I'd be glad to help.

**Comment by Madhu-1** [comment]
@mehdy we got the required support for rbd snapshot and clone in master now, Can you please resolve the merge conflict on this one no

**Comment by Madhu-1** [comment]
@dillaman can you please review this one.

**Comment by Madhu-1** [comment]
@mehdy Marking it for 3.1.0. let me know if you need any help

**Comment by mehdy** [comment]
@Madhu-1 I haven't had much time lately but I'll try to make time for it. Thanks.

**Comment by mehdy** [comment]
It's ready to be reviewed.
@dillaman Would you please review it?

File: internal/util/csiconfig.go (hunk 5 above)
Comment (nixpanic) [reply internal/util/csiconfig.go]: this comment still mentions `"namespace"`, not `"radosNamespace"` as it is used everywhere else

Comment (nixpanic) [reply internal/util/csiconfig.go]: ah, it gets addressed in the next commit :+1:

File: internal/rbd/controllerserver.go [hunk 7]
```diff
@@ -145,6 +148,7 @@ func buildCreateVolumeResponse(ctx context.Context, req *csi.CreateVolumeRequest
 	volumeContext := req.GetParameters()
 	volumeContext["pool"] = rbdVol.Pool
 	volumeContext["journalPool"] = rbdVol.JournalPool
+	volumeContext["radosNamespace"] = rbdVol.RadosNamespace
```
[the remark this answers is not in the document]
Comment (mehdy) [reply internal/rbd/controllerserver.go:151]: sure!

File: internal/rbd/rbd_util.go [hunk 8]
```diff
@@ -86,9 +86,11 @@ type rbdVolume struct {
 	JournalPool         string
 	Pool                string `json:"pool"`
 	DataPool            string
+	RadosNamespace      string
 	ImageID             string
 	ParentName          string
 	imageFeatureSet     librbd.FeatureSet
+	ImageFeatures       string `json:"imageFeatures"`
```
[the remark this answers is not in the document]
Comment (mehdy) [reply internal/rbd/rbd_util.go:93]: Not at all. I must've made a mistake when I was resolving the conflicts from the recent changes.

File: internal/cephfs/fsjournal.go [hunk 9]
```diff
@@ -53,7 +53,7 @@ func checkVolExists(ctx context.Context, volOptions *volumeOptions, secret map[s
 	}
 	defer cr.DeleteCredentials()
 
-	j, err := volJournal.Connect(volOptions.Monitors, cr)
+	j, err := volJournal.Connect(volOptions.Monitors, radosNamespace, cr)
```
[the remark this answers is not in the document]
Comment (mehdy) [reply internal/cephfs/fsjournal.go:56]: good point. will do.

File: examples/csi-config-map-sample.yaml [hunk 10]
```diff
@@ -12,6 +12,10 @@ kind: ConfigMap
 # each such cluster in use.
 # To add more clusters or edit MON addresses in an existing configmap, use
 # the `kubectl replace` command.
+# The <rados-namespace> is optional and represents a namespace in the pool.
+# If any given all the of the rbd images, snapshots and other metadata
+# will be store within the namespace.
+# NOTE: The given namespace must already exists in the pool.
```
[the remark this answers is not in the document]
Comment (mehdy) [reply examples/csi-config-map-sample.yaml:18]: Absolutely!

**Comment by mehdy** [comment]
@Madhu-1 Thanks for the review. made some changes.

**Comment by mehdy** [comment]
@Madhu-1 Resolved the conflicts and ready to merged.

File: examples/csi-config-map-sample.yaml [hunk 11]
```diff
@@ -12,6 +12,12 @@ kind: ConfigMap
 # each such cluster in use.
 # To add more clusters or edit MON addresses in an existing configmap, use
 # the `kubectl replace` command.
+# The <rados-namespace> is optional and represents a radosNamespace in the pool.
+# If any given all the of the rbd images, snapshots and other metadata
```
Comment (humblec) [inline examples/csi-config-map-sample.yaml:16]: to me, it feels like we need to reword this and below line @mehdy

**Comment by humblec** [comment]
@mehdy have we opened an issue for this or is it already resolved ?
```
When running e2e tests there was an error with the raw-block-pod but testing it again without any of changes against a clean environment the issue still exists (maybe it's another issue).
```

**Comment by mehdy** [comment]
> @mehdy have we opened an issue for this or is it already resolved ?
> 
> ```
> When running e2e tests there was an error with the raw-block-pod but testing it again without any of changes against a clean environment the issue still exists (maybe it's another issue).
> ```

This was an issue with my dev environment.

**Comment by Madhu-1** [comment]
@humblec do you have any final comments on this feature?  Can we take this in for v3.1.0?
Any code changes required?

File: examples/csi-config-map-sample.yaml (hunk 11 above)
Comment (mehdy) [reply examples/csi-config-map-sample.yaml:16]: Sure! How about this?

"To use a specific rados namespace for your rbd images and snapshots, provide the name of the desired rados namespace".

If you have something else in mind, I'd be glad to use the help.

Comment (humblec) [reply examples/csi-config-map-sample.yaml:16]: @mehdy it looks good to me ! :+1:

**Comment by humblec** [comment]
> > @mehdy have we opened an issue for this or is it already resolved ?
> > ```
> > When running e2e tests there was an error with the raw-block-pod but testing it again without any of changes against a clean environment the issue still exists (maybe it's another issue).
> > ```
> 
> This was an issue with my dev environment.

@mehdy can you removing it from the PR description then?

**Comment by mehdy** [comment]
@humblec Done :+1:

**Comment by humblec** [comment]
> @humblec do you have any final comments on this feature? Can we take this in for v3.1.0?
> Any code changes required?

@madhu-1 as discussed, lets take this in for 3.1.0, especially this was deferred from 3.0.0 to avoid last minute hiccups with 3.0.0 release!

**Comment by humblec** [comment]
> @humblec Done +1

@mehdy Thanks !! If at ll possible, its also good to have a mention or documentation around this change. But it can be a followup PR or can be discarded too!

**Comment by Madhu-1** [comment]
> > @humblec do you have any final comments on this feature? Can we take this in for v3.1.0?
> > Any code changes required?
> 
> @Madhu-1 as discussed, lets take this in for 3.1.0, especially this was deferred from 3.0.0 to avoid last minute hiccups with 3.0.0 release!
As we have minimal upgrade testing

@humblec I would suggest testing the backward compatibility before taking it in.

**Comment by mehdy** [comment]
> > @humblec Done +1
> 
> @mehdy Thanks !! If at ll possible, its also good to have a mention or documentation around this change. But it can be a followup PR or can be discarded too!

@humblec You're welcome! Sure. I'll see what I can do and I'll open a new PR if I had something to add.

**Comment by humblec** [comment]
Merging this manually as discussed in the Slack as the auto merge is blocked for long time.
